R/zchunk_LA123.electricity.R

```diff
@@ -29,26 +29,118 @@ module_energy_LA123.electricity_DISABLED <- function(command, ...) {
+    # based on electricity inputs and outputs by fuel and year.
+    L123.in_EJ_R_elec_F_Yh%>%
+      gather(X_HISTORICAL_YEARS,input,-GCAM_region_ID,-sector,-fuel)->input_fuels_by_year_region
+
+    # Calculates electricity generation efficiencies (L123.eff_R_elec_F_Yh) by fuel and year
+    L123.out_EJ_R_elec_F_Yh%>%
+      semi_join(L123.in_EJ_R_elec_F_Yh,by=c("GCAM_region_ID", "fuel"))%>%
+      gather(X_HISTORICAL_YEARS,outputs,-GCAM_region_ID,-sector,-fuel)%>%
+      left_join(input_fuels_by_year_region, by=c("GCAM_region_ID","sector","fuel","X_HISTORICAL_YEARS"))%>%
+      mutate(eff=outputs/input)%>%
+      select(-outputs,-input)%>%
+      spread(X_HISTORICAL_YEARS,eff)->L123.eff_R_elec_F_Yh
+
+    # Taking care of NA, 0, and INF values generatted in previous step (efficiency calculations) and updating
+    # the efficiency output L123.eff_R_elec_F_Yh
+    L123.eff_R_elec_F_Yh%>%
+      mutate_at(funs(replace(., is.na(.), DEFAULT_ELECTRIC_EFFICIENCY)), .cols = X_HISTORICAL_YEARS)%>%
+      mutate_at(funs(replace(., .==0, DEFAULT_ELECTRIC_EFFICIENCY)), .cols = X_HISTORICAL_YEARS)%>%
+      mutate_at(funs(replace(., .==Inf, DEFAULT_ELECTRIC_EFFICIENCY)), .cols = X_HISTORICAL_YEARS)->L123.eff_R_elec_F_Yh
+
+    # Tibble (Output_efficiency_based) created to adjust electricity outputs (L123.out_EJ_R_elec_F_Yh)
+    # based on above modified efficiency calculations
+    L123.in_EJ_R_elec_F_Yh%>%
+      gather(X_HISTORICAL_YEARS,outputs,-GCAM_region_ID,-sector,-fuel)%>%
+      left_join(gather(L123.eff_R_elec_F_Yh,X_HISTORICAL_YEARS,outputs,-GCAM_region_ID,-sector,-fuel),
+                by=c("GCAM_region_ID","sector","fuel","X_HISTORICAL_YEARS"))%>%
+      mutate(outputs_new=outputs.x*outputs.y)%>%
+      select(-outputs.x,-outputs.y)->Output_efficiency_based
+
+    # Adjusting electricity outputs based on new efficiencies
+    L123.out_EJ_R_elec_F_Yh%>%
+      gather(X_HISTORICAL_YEARS,outputs,-GCAM_region_ID,-sector,-fuel)%>%
+      left_join(Output_efficiency_based,by=c("GCAM_region_ID","sector","fuel","X_HISTORICAL_YEARS"))%>%
+      mutate(outputs =if_else(!is.na(outputs_new),outputs_new,outputs))%>%
+      select(-outputs_new)%>%
+      spread(X_HISTORICAL_YEARS,outputs)->L123.out_EJ_R_elec_F_Yh
+
+    #2: CHP calculations
+
+    # filter end use fuel in the industry sector and remove any NA generated.
+    enduse_fuel_aggregation%>%
+      select(fuel,industry)%>%
+      filter(!is.na(industry))->enduse_fuel_aggregation_industry
+
+    # Create Industry CHP estimates by fuel, region and industry sector.
+    L1011.en_bal_EJ_R_Si_Fi_Yh%>%
+      filter(sector == "out_chp_elec")%>%
+      mutate(sector = "chp_elec")%>%
+      left_join(enduse_fuel_aggregation_industry,by="fuel")%>%
+      mutate(fuel=industry)%>%
+      select(-industry)%>%
+      mutate(fuel= replace(fuel,fuel== "heat",NA))%>%
+      group_by(GCAM_region_ID, sector,fuel)%>%
```

Comment:
This will become (I think) `group_by(GCAM_region_ID, sector, fuel, year) %>% summarise(value = sum(value))`